src/solvers/refinement/string_constraint_generator_valueof.cpp

```diff
@@ -313,102 +296,83 @@ string_exprt string_constraint_generatort::add_axioms_from_int(
   exprt max=from_integer(max_size, index_type);
 
   // We add axioms:
-  // a1 : 0 <|res|<=max_size
-  // a2 : (res[0]='-')||('0'<=res[0]<='9')
-
-  and_exprt a1(res.axiom_for_is_strictly_longer_than(zero),
-               res.axiom_for_is_shorter_than(max));
+  // a1 : i < 0 => 1 <|res|<=max_size && res[0]='-'
+  // a2 : i >= 0 => 0 <|res|<=max_size-1 && '0'<=res[0]<='9'
+  // a3 : |res|>1 && '0'<=res[0]<='9' => res[0]!='0'
+  // a4 : |res|>1 && res[0]='-' => res[1]!='0'
+
+  binary_relation_exprt is_negative(i, ID_lt, zero);
+  and_exprt correct_length1(res.axiom_for_is_strictly_longer_than(1),
+                            res.axiom_for_is_shorter_than(max));
+  equal_exprt starts_with_minus(res[0], minus_char);
+  implies_exprt a1(is_negative, and_exprt(correct_length1, starts_with_minus));
   axioms.push_back(a1);
 
-  exprt chr=res[0];
-  equal_exprt starts_with_minus(chr, minus_char);
-  and_exprt starts_with_digit(
-    binary_relation_exprt(chr, ID_ge, zero_char),
-    binary_relation_exprt(chr, ID_le, nine_char));
-  or_exprt a2(starts_with_digit, starts_with_minus);
+  not_exprt is_positive(is_negative);
+  and_exprt starts_with_digit(binary_relation_exprt(res[0], ID_ge, zero_char),
+                              binary_relation_exprt(res[0], ID_le, nine_char));
+  and_exprt correct_length2(res.axiom_for_is_strictly_longer_than(0),
+                            res.axiom_for_is_strictly_shorter_than(max));
+  implies_exprt a2(is_positive, and_exprt(correct_length2, starts_with_digit));
   axioms.push_back(a2);
 
-  // These are constraints to detect number that requiere the maximum number
-  // of digits
-  exprt smallest_with_max_digits=
-    from_integer(smallest_by_digit(max_size-1), type);
-  binary_relation_exprt big_negative(
-    i, ID_le, unary_minus_exprt(smallest_with_max_digits));
-  binary_relation_exprt big_positive(i, ID_ge, smallest_with_max_digits);
-  or_exprt requieres_max_digits(big_negative, big_positive);
+  implies_exprt a3(
+    and_exprt(res.axiom_for_is_strictly_longer_than(1), starts_with_digit),
+    not_exprt(equal_exprt(res[0], zero_char)));
+  axioms.push_back(a3);
+
+  implies_exprt a4(
+    and_exprt(res.axiom_for_is_strictly_longer_than(1), starts_with_minus),
+    not_exprt(equal_exprt(res[1], zero_char)));
+  axioms.push_back(a4);
 
   for(size_t size=1; size<=max_size; size++)
```

Comment:
Do you check whether max_size is strictly smaller than the largest value of size_t? Probably not a practical problem, but otherwise this might loop infinitely.